include/material/modified_cam_clay.tcc

```diff
@@ -0,0 +1,722 @@
+    df_dpc *= subloading_r;
+    df_dpcd *= subloading_r;
+    df_dpcc = p - subloading_r * (p + pc + pcd + 2 * pcc);
+  }
+  // Upsilon
+  const double upsilon =
+      (1 + (*state_vars).at("void_ratio")) / (lambda_ - kappa_);
+  // Coefficients in plastic stiffness matrix
+  const double a1 = pow(((*state_vars).at("bulk_modulus") * df_dp), 2);
+  const double a2 = -sqrt(6) * (*state_vars).at("bulk_modulus") * df_dp *
+                    (*state_vars).at("shear_modulus") * df_dq;
+  const double a3 = 6 * pow(((*state_vars).at("shear_modulus") * df_dq), 2);
+  // Numerator
+  double num = (*state_vars).at("bulk_modulus") * (df_dp * df_dp) +
+               3 * (*state_vars).at("shear_modulus") * (df_dq * df_dq);
+
+  // Hardening parameter
+  double hardening = upsilon * pc * df_dp * df_dpc;
+  // Compute bonding
+  if (bonding_) {
+    // Compute pcd hardening parameter
+    const double hardening_pcd =
+        df_dpcd * (-m_degradation_ * mc_b_ * pcd) * df_dq;
+    // Compute pcc hardening parameter
+    const double hardening_pcc =
+        df_dpcc * (-m_degradation_ * mc_d_ * pcc) * df_dq;
+    // Update hardening parameter
+    hardening += (hardening_pcd + hardening_pcc);
+  }
+  // Compute subloading
+  if (subloading_) {
+    // Compute dF / dR
+    const double df_dr = -(p + pcc) * (pc + pcd + pcc);
+    // Compute subloading hardening parameter
+    const double hardening_subloading =
+        -df_dr * subloading_u_ * (1 + (pcd + pcc) / pc) * log(subloading_r) *
+        sqrt(std::pow((*state_vars).at("dpvstrain"), 2) +
+             std::pow((*state_vars).at("dpdstrain"), 2));
+    // Update hardening parameter
+    hardening += hardening_subloading;
+  }
+  // Compute the deviatoric stress
+  Vector6d dev_stress = Vector6d::Zero();
+  dev_stress(0) = stress(0) + p;
+  dev_stress(1) = stress(1) + p;
+  dev_stress(2) = stress(2) + p;
+  dev_stress(3) = stress(3);
+  if (Tdim == 3) {
+    dev_stress(4) = stress(4);
+    dev_stress(5) = stress(5);
+  }
+  Matrix6x6 n_l = Matrix6x6::Zero();
+  Matrix6x6 l_n = Matrix6x6::Zero();
+  Matrix6x6 l_l = Matrix6x6::Zero();
+  Matrix6x6 n_n = Matrix6x6::Zero();
+  double xi = q / sqrt(1.5);
+  // lxn
+  if (xi > std::numeric_limits<double>::epsilon()) {
+    for (int i = 0; i < 3; ++i) {
+      for (int j = 0; j < 6; ++j) l_n(i, j) = dev_stress(j) / xi;
+    }
+    // nxl
+    for (int i = 0; i < 6; ++i) {
+      for (int j = 0; j < 3; ++j) n_l(i, j) = dev_stress(i) / xi;
+    }
+    // nxn
+    for (int i = 0; i < 6; ++i) {
+      for (int j = 0; j < 6; ++j) {
+        n_n(i, j) = dev_stress(i) * dev_stress(j) / (xi * xi);
+      }
+    }
+    // lxl
+    for (int i = 0; i < 3; ++i) {
+      for (int j = 0; j < 3; ++j) l_l(i, j) = 1;
+    }
+  }
+  // Compute plastic tensor
+  this->dp_ = (a1 * l_l + a2 * (n_l + l_n) + a3 * (n_n)) / (num - hardening);
+
+  return true;
+}
+
+//! Compute stress invariants
+template <unsigned Tdim>
+bool mpm::ModifiedCamClay<Tdim>::compute_stress_invariants(
```

Comment:
I think as al the constitutive models are merged, it is good to refactor all the common functions like this in the next PR @kks32 @ezrayst